src/application/estimator/estimator.c

```diff
@@ -1 +1,75 @@
 #include "application/estimator/estimator.h"
+#include "FreeRTOS.h"
+#include "application/controller/controller.h"
+#include "stm32h7xx_hal.h"
+#include <queue.h>
+#include <semphr.h>
+
+QueueHandle_t imu_data_handle = NULL;
+estimator_imu_measurement_t imu_data_buffer;
+const int STATE_EST_CAN_RATE = 3;
+uint8_t state_est_can_counter = 0;
+controller_output_t controller_input_buffer;
+controller_input_t output_to_controller;
+
+w_status_t estimator_init() {
+    imu_data_handle = xQueueCreate(1, sizeof(imu_data_buffer));
+
+    if (imu_data_handle == NULL) {
+        return W_FAILURE;
+    }
+    return W_SUCCESS;
+}
+
+w_status_t estimator_initialize_filter(void) {
+    // check for message from can_handler
+
+    // initialize filter
+
+    return W_SUCCESS;
+}
+
+w_status_t estimator_update_inputs_imu(estimator_all_imus_input_t *data) {
+    if (data == NULL) {
+        return W_FAILURE;
+    }
+    if (xQueueOverwrite(imu_data_handle, data) != pdPASS) {
+        return W_FAILURE;
+    }
+    return W_SUCCESS;
+}
+
+void estimator_task(void *argument) {
+    /*bool filter_is_initialized = false;
+    while (filter_is_initialized != true){
+        // check for estimator initialization by CAN
+    }*/
+    (void)argument;
+
+    if (xQueueReceive(imu_data_handle, &imu_data_buffer, pdMS_TO_TICKS(5)) != pdTRUE) {
+        // log_text("State estimation", "failed to receive imu data within 5ms");
+        return;
+    }
+
+    if (controller_get_latest_output(&controller_input_buffer) != W_SUCCESS) {
+        // log_text("State estimation", "failed to receive controller data");
+        return;
+    }
+
+    // double x_new[13];
+    // double P_new[13 * 13];
+
+    // ekf_algorithm(x_new, P_new, imu_data_buffer, controller_input_buffer);
+
+    // write information from x_new and P_new into output_to_controller
+
+    if (controller_update_inputs(&output_to_controller) != W_SUCCESS) {
+        // log_text("State estimation", "failed to give controller the output from state
+        // estimation");
+    }
+
+    if (state_est_can_counter % STATE_EST_CAN_RATE == 0) {
+        // send to CAN
+    }
+    ++state_est_can_counter;
+}
```

Comment:
you need to wrap your task function in an infinite loop and then delay until 5 ms pass. You should define that 5 ms value as smth like `STATE_EST_TASK_DELAY_MS`